tests/integration/telemetry/wasmplugin/wasmplugin_test.go

```diff
@@ -0,0 +1,375 @@
+		}
+
+		if successCount < 2 {
+			successCount++
+			return fmt.Errorf("retry")
+		}
+		return nil
+	}, retry.Delay(1*time.Second), retry.Timeout(100*time.Second))
+}
+
+func TestWasmPluginPullPolicy(t *testing.T) {
+	framework.NewTest(t).
+		Features("extensibility.wasm.remote-load").
+		Run(func(t framework.TestContext) {
+			applyAndTestWasm(t, wasmTestConfigs{
+				desc:            "initial creation with latest",
+				name:            "wasm-test-module",
+				tag:             "latest",
+				policy:          "",
+				upstreamVersion: "0.0.1",
+				expectedVersion: "0.0.1",
+			})
+
+			resetWasm(t, "wasm-test-module")
+			applyAndTestWasm(t, wasmTestConfigs{
+				desc:            "upstream is upgraded to 0.0.2. 0.0.1 is already present and policy is IfNotPresent, so should not pull",
+				name:            "wasm-test-module",
+				tag:             "latest",
+				policy:          "IfNotPresent",
+				upstreamVersion: "0.0.2",
+				expectedVersion: "0.0.1",
+			})
+
+			// Intentionally, do not reset here to see the upgrade from 0.0.1.
+			applyAndTestWasm(t, wasmTestConfigs{
+				desc:            "upstream is upgraded to 0.0.2. 0.0.1 is already present. But policy is default and tag is latest, so pull the image",
+				name:            "wasm-test-module",
+				tag:             "latest",
+				policy:          "",
+				upstreamVersion: "0.0.2",
+				expectedVersion: "0.0.2",
+			})
+			resetWasm(t, "wasm-test-module")
+
+			applyAndTestWasm(t, wasmTestConfigs{
+				desc:            "initial creation with 0.0.1",
+				name:            "wasm-test-module-test-tag-1",
+				tag:             "test-tag-1",
+				policy:          "",
+				upstreamVersion: "0.0.1",
+				expectedVersion: "0.0.1",
+			})
+
+			resetWasm(t, "wasm-test-module-test-tag-1")
+			applyAndTestWasm(t, wasmTestConfigs{
+				desc:            "upstream is upgraded to 0.0.2, but 0.0.1 is already present and policy is IfNotPresent",
+				name:            "wasm-test-module-test-tag-1",
+				tag:             "test-tag-1",
+				policy:          "IfNotPresent",
+				upstreamVersion: "0.0.2",
+				expectedVersion: "0.0.1",
+			})
+
+			resetWasm(t, "wasm-test-module-test-tag-1")
+			applyAndTestWasm(t, wasmTestConfigs{
+				desc:            "upstream is upgraded to 0.0.2, but 0.0.1 is already present and policy is default",
+				name:            "wasm-test-module-test-tag-1",
+				tag:             "test-tag-1",
+				policy:          "",
+				upstreamVersion: "0.0.2",
+				expectedVersion: "0.0.1",
+			})
+
+			// Intentionally, do not reset here to see the upgrade from 0.0.1.
+			applyAndTestWasm(t, wasmTestConfigs{
+				desc:            "upstream is upgraded to 0.0.2. 0.0.1 is already present but policy is Always, so pull 0.0.2",
+				name:            "wasm-test-module-test-tag-1",
+				tag:             "test-tag-1",
+				policy:          "Always",
+				upstreamVersion: "0.0.2",
+				expectedVersion: "0.0.2",
+			})
+		})
+}
+
+func TestMain(m *testing.M) {
+	framework.NewSuite(m).
```

Comment:
this will create a new suite?

Reply:
yes. 

Could you explain your thoughts in detail if you think another way?

Reply:
https://github.com/istio/istio/pull/36677#pullrequestreview-842312449

Reply:
Moved to stats/prometheus/wasm and share the suite.